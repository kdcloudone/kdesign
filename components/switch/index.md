---
category: Components
type: 录入
order: 14
title: Switch
subtitle: 开关
---


开关是一种用于两种互斥状态之间进行切换的组件。

## 使用场景
需要触发激活或停用某些功能时，例如：“启用”和“不启用”。

## API

| 参数              | 说明                              | 类型                                     | 默认值 | 版本 |
| ----------------- | --------------------------------- | ---------------------------------------- | ------ | ---- |
| className         | Switch 器类名                     | string                                   | -      |      |
| checked           | 指定当前是否选中                  | boolean                                  | false  |      |
| checkedChildren   | 选中时的内容                      | string\|ReactNode                        |        |      |
| defaultChecked    | 初始是否选中                      | boolean                                  | false  |      |
| disabled          | 是否禁用                          | boolean                                  | false  |      |
| loading           | 加载中的开关                      | boolean                                  | false  |      |
| size              | 开关大小，可选值：`large` `small` | string                                   | small  |      |
| unCheckedChildren | 非选中时的内容                    | string\|ReactNode                        |        |      |
| onClick           | 点击时回调函数                    | Function(checked: boolean, event: Event) |        |      |
| onChange          | 变化时回调函数                    | Function(checked: boolean, event: Event)               |        |      |
