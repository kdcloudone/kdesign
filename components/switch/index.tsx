import Switch from './switch'
export { ISwitchProps } from './switch'
export default Switch
