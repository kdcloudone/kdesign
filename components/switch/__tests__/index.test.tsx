import React from 'react'
import { render, mount } from 'enzyme'
import mountTest from '../../../tests/shared/mountTest'
import Switch from '../index'
import { SwitchSizes } from '../switch'

describe('Switch', () => {
  mountTest(Switch)

  it('renders correctly', () => {
    expect(render(<Switch />)).toMatchSnapshot()
    SwitchSizes.forEach((size) => {
      expect(render(<Switch size={size} />)).toMatchSnapshot()
    })
  })

  it('should have displayName static property', () => {
    const wrapper = mount(<Switch />)
    expect((wrapper.type() as any).displayName).toBe('Switch')
  })

  it('should class use right', () => {
    const DefaultSwitch = mount(<Switch />)
    expect(DefaultSwitch.find('.kd-switch')).toHaveClassName('.kd-switch-size-small')

    const DefaultCheckedSwitch = mount(<Switch defaultChecked />)
    expect(DefaultCheckedSwitch.find('.kd-switch')).toHaveClassName('.kd-switch-checked')

    const CheckedSwitch = mount(<Switch checked={true} />)
    expect(CheckedSwitch.find('.kd-switch')).toHaveClassName('.kd-switch-checked')

    const LoadingSwitch = mount(<Switch loading defaultChecked />)
    expect(LoadingSwitch.find('.kd-switch')).toHaveClassName('.kd-switch-loading')
    expect(LoadingSwitch.find('.kd-switch')).toHaveClassName('.kd-switch-disabled')

    const DisabledSwitch = mount(<Switch disabled />)
    expect(DisabledSwitch.find('.kd-switch')).toHaveClassName('.kd-switch-disabled')
  })

  it('should not clickable when switch is loading', () => {
    const onClick = jest.fn()
    const wrapper = mount(<Switch loading onClick={onClick} />)
    wrapper.simulate('click')
    expect(onClick).not.toHaveBeenCalledWith()
  })

  it('should not clickable when switch is disabled', () => {
    const onClick = jest.fn()
    const wrapper = mount(<Switch disabled onClick={onClick} />)
    wrapper.simulate('click')
    expect(onClick).not.toHaveBeenCalledWith()
  })

  it('should clickable', () => {
    const onClick = jest.fn()
    const wrapper = mount(<Switch onClick={onClick} />)
    wrapper.simulate('click')
    expect(onClick).toHaveBeenCalled()
  })

  it('should onChange been emit when click', () => {
    const onChange = jest.fn()
    const wrapper = mount(<Switch onChange={onChange} />)
    wrapper.simulate('click')
    expect(onChange).toHaveBeenCalled()
  })
})
