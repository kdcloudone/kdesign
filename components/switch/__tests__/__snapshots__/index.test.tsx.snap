// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch renders correctly 1`] = `
<span
  class="kd-switch kd-switch-size-small"
>
  <div
    class="kd-switch-inner"
  />
</span>
`;

exports[`Switch renders correctly 2`] = `
<span
  class="kd-switch kd-switch-size-large"
>
  <div
    class="kd-switch-inner"
  />
</span>
`;

exports[`Switch renders correctly 3`] = `
<span
  class="kd-switch kd-switch-size-small"
>
  <div
    class="kd-switch-inner"
  />
</span>
`;
